pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -797,6 +801,18 @@ private void addNewRow(int docId, GenericRow row) {
             recordIndexingError(indexEntry.getKey(), e);
           }
         }
+
+        if (_thresholdForNumOfColValuesEnabled) {
+          int prevCount = indexContainer._valuesInfo.getNumValues();
+          long newCount = prevCount + 1L + values.length;
```

Comment:
Total values itself is not enough. We should perform a per-index check (add an api to the `MutableIndex` and let it return if it can take more values).
E.g. for MV forward index, if we get 1B values, but each value takes more than 2 bytes, we will run into the same exception

Reply:
Mutable index is unbounded right (No code enforced limit)? From the code I see Realtime Mutable index is always created with dictionary (even for MV VarByte col with noDict enabled in config). 

Hence from `numOfValues` we know that size of mutable index is approx: `numOfValues * 4 Bytes`. 

But we are more interested in the size of immutable index since that's where exception is being thrown. However Immutable index can be larger or even smaller than mutable index as implementation is completely diff. So while building mutable index we need to keep some state to estimate the approx size of the immutable version of the index (Like while building mutable fwd index, we need to keep an estimation of bitmap, numBitsPerValue, header size, etc).

> We should perform a per-index check (add an api to the MutableIndex and let it return if it can take more values).

So this might be adding too much complexity to every mutable index since now after every row consumption we need to update estimated size of the corresponding immutable index and it tight couples code with immutable index logic.

Reply:
hmm. Okay, let me figure out if we can even do this on ingestion side considering we have 5 different implementations of immutable forward index for MV columns each having diff size limit.

Reply:
We can keep threshold near 500 million for numOfValues which will be good enough for few immutable indexes but not for varByte fwd index having 4GB Limit.

Reply:
For now we can figure out a conservative value that will never cause over 2GB buffer. Once we support larger index, we can increate this value

Reply:
Sure. Will pick the task to keep the indexes unbounded post this.